File: src/chapters/minimization/content/code/golden.go
package main

import (
	"fmt"
	"math"
)

// Golden-section search for the minimum of a unimodal f on [a, b].
func goldenSection(f func(float64) float64, a, b, tol float64) float64 {
	g := (math.Sqrt(5) - 1) / 2
	c, d := b-g*(b-a), a+g*(b-a)
	fc, fd := f(c), f(d)
	for b-a > tol {
		if fc < fd {
			b, d, fd = d, c, fc
			c = b - g*(b-a)
			fc = f(c)
		} else {
			a, c, fc = c, d, fd
			d = a + g*(b-a)
			fd = f(d)
		}
	}
	return (a + b) / 2
}

func main() {
	fmt.Println(goldenSection(func(x float64) float64 { return (x-2)*(x-2) + 1 }, 0, 5, 1e-8))
}

File: src/chapters/minimization/content/code/golden.r
# Golden-section search for the minimum of a unimodal f on [a, b].
golden_section <- function(f, a, b, tol = 1e-8) {
  g <- (sqrt(5) - 1) / 2
  c <- b - g * (b - a)
  d <- a + g * (b - a)
  fc <- f(c)
  fd <- f(d)
  while (b - a > tol) {
    if (fc < fd) {
      b <- d; d <- c; fd <- fc
      c <- b - g * (b - a); fc <- f(c)
    } else {
      a <- c; c <- d; fc <- fd
      d <- a + g * (b - a); fd <- f(d)
    }
  }
  (a + b) / 2
}

if (sys.nframe() == 0) {
  print(golden_section(function(x) (x - 2)^2 + 1, 0, 5))   # -> 2
}
